# Exploratory Data Analysis, Hop Teaming Dataset

### Import libraries, connect to the database, and take a look around

In [1]:
# import libraries
import pandas as pd
import sqlite3
from tqdm.notebook import tqdm

In [2]:
# Set the display to show more rows and columns
pd.options.display.max_rows = 500
pd.options.display.max_columns = 200

In [3]:
# Assign sqlite database to db variable
db = sqlite3.connect('../data/hop_teaming.sqlite')

In [4]:
# Check the tables in the database
tables = db.execute("SELECT name FROM sqlite_master WHERE type='table';")
for table in tables:
    print(table[0])

cbsa
referrals
nashville_referrals
nppes
specialty


In [5]:
# For reference: pull out the column names for a given table
col = db.execute("PRAGMA table_info(specialty)").fetchall()
print([e[1] for e in col])

['npi', 'primary_taxonomy', 'Classification']


### How many providers are in the nppes dataset?
*Note that `entity_type_code` 1 = Providers and `entity_type_code` 2 = Facilities.*

In [6]:
all_providers = """
SELECT COUNT(npi), COUNT(DISTINCT npi)
FROM nppes
"""
all_providers = pd.read_sql(all_providers, db)

In [7]:
all_providers

COUNT(npi)  COUNT(DISTINCT npi)
0     6714038              6714038

In [8]:
# Create a query to get all providers with a practice location within the Nashville CBSA
nash_providers = """
    SELECT n.*

    FROM nppes AS n

    JOIN cbsa AS c
        ON c.ZIP = CAST(n.zip_5 AS INTEGER)
"""

# Read the query results into a dataframe
nash_nppes = pd.read_sql(nash_providers, db)

In [9]:
# Take a look at the number of datapoints available
nash_nppes.info()
# There are 38,414 rows in the table

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38414 entries, 0 to 38413
Data columns (total 15 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   npi                         38414 non-null  int64  
 1   entity_type_code            38414 non-null  float64
 2   provider_organization_name  7574 non-null   object 
 3   provider_last_name          30839 non-null  object 
 4   provider_first_name         30840 non-null  object 
 5   provider_middle_name        19914 non-null  object 
 6   provider_name_prefix        11560 non-null  object 
 7   provider_name_suffix        907 non-null    object 
 8   provider_credential         25323 non-null  object 
 9   address_01                  38414 non-null  object 
 10  address_02                  9957 non-null   object 
 11  city                        38414 non-null  object 
 12  state                       38414 non-null  object 
 13  zip_9                       384

In [10]:
# Take a look at the unique provider credential types
nash_nppes.provider_credential.unique().tolist()

['CRNP',
 'M.D.',
 'MD',
 'FNP',
 None,
 'M. D.',
 'DO',
 'PHARMD',
 'PA',
 'D.M.D.',
 'RN, APN',
 'DDS',
 'CRNA, MS',
 'NP',
 'APN',
 'APRN',
 'C.R.N.A., A.P.N.',
 'FNP-C',
 'D.C.',
 'RPH, PHARMD,MS',
 'O.D.',
 'P.T., D.P.T',
 'D.D.S.',
 'N.P.',
 'APRN, BC, FNP',
 'PAC',
 'ACNP',
 'M.S.S.W.',
 'GNP',
 'MD, PHD',
 'MSN APRN-BC',
 'DPM',
 'PHARM. D.',
 'PH.D.',
 'CRNA',
 'NP-C',
 'PA-C',
 'PHARM.D.',
 'APRN, BC',
 'O.D',
 'M.D., PH.D.',
 'D.D.S',
 'LCSW',
 'LCSW, ACSW',
 'L.C.S.W.',
 'DDS, MD',
 'DO, MBA',
 'DC',
 'M.D., R.V.T',
 'ARNP',
 'PT',
 'D.O. F.A.C.O.G.',
 'DMD',
 'M.S., P.A.-C.',
 'D.O.',
 'MS, PT',
 'P.A-C..',
 'CNM',
 'MSPT',
 'DO, RDMS',
 'P.T.',
 'PHD CCC-SLP',
 'DPT',
 'DP',
 'D.P.T.',
 'PTA',
 'COTA',
 'O.T.',
 'D.PH.',
 'PHD, FNP- BC',
 'DNP,FNP- BC',
 'MD MBA',
 'D.P.M.',
 'ACNP-BC',
 'LDN',
 'MSN, APRN-BC, CNN',
 'DNP',
 'OD',
 'FNP C',
 'ED.D.',
 'ANPC',
 'MSSW, LCSW',
 'MD, MS, FIPP',
 'M.D., MPH',
 'DDS,PC',
 'WHNP CNM',
 'MSN',
 'CFNP',
 'PSY.D.',
 'ANP-C',
 'P.A.

*The credentials are VERY not-standardized and include multiple variants of the same thing with different spellings (e.g. `PharmD` vs. `Pharm.D.` vs. `PHARMD`) in addition to commas within the field itself (e.g. `APN-BC, NP-C`). If we need to use this field, we will need to do some data cleaning.*

In [11]:
# Subset to only individual providers
ind_nash_nppes = nash_nppes[nash_nppes['entity_type_code'] == 1]

# Take a look at how many individual providers there are in the Nashville area
ind_nash_nppes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30840 entries, 0 to 38413
Data columns (total 15 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   npi                         30840 non-null  int64  
 1   entity_type_code            30840 non-null  float64
 2   provider_organization_name  0 non-null      object 
 3   provider_last_name          30839 non-null  object 
 4   provider_first_name         30840 non-null  object 
 5   provider_middle_name        19914 non-null  object 
 6   provider_name_prefix        11560 non-null  object 
 7   provider_name_suffix        907 non-null    object 
 8   provider_credential         25323 non-null  object 
 9   address_01                  30840 non-null  object 
 10  address_02                  7556 non-null   object 
 11  city                        30840 non-null  object 
 12  state                       30840 non-null  object 
 13  zip_9                       308

In [12]:
# Who is the provider without a last name?
ind_nash_nppes[ind_nash_nppes['provider_last_name'].isna() == True]

npi  entity_type_code provider_organization_name  \
10346  1215064852               1.0                       None   

      provider_last_name provider_first_name provider_middle_name  \
10346               None           ELIZABETH                  ANN   

      provider_name_prefix provider_name_suffix provider_credential  \
10346                  DR.                 None                M.D.   

              address_01 address_02      city state        zip_9  zip_5  
10346  1222 TROTWOOD AVE  SUITE 101  COLUMBIA    TN  384016436.0  38401

*Hilariously, [this is a real provider](https://npiregistry.cms.hhs.gov/registry/provider-view/1215064852) whose last name is "Null".*

### Explore the specialties data

In [13]:
specialties_raw = """
    SELECT *
    FROM specialty
"""
specialties_raw = pd.read_sql(specialties_raw, db)

In [14]:
specialties_raw

npi primary_taxonomy  \
0        1679576722       207X00000X   
1        1588667638       207RC0000X   
2        1497758544       251G00000X   
3        1306849450       2085R0202X   
4        1215930367       207RH0003X   
...             ...              ...   
6714033  1326630724       363LF0000X   
6714034  1912599325       183500000X   
6714035  1821680232       163WP0808X   
6714036  1730771148       390200000X   
6714037  1649862053       235Z00000X   

                                            Classification  
0                                      Orthopaedic Surgery  
1                                        Internal Medicine  
2                            Hospice Care, Community Based  
3                                                Radiology  
4                                        Internal Medicine  
...                                                    ...  
6714033                                 Nurse Practitioner  
6714034                                         Pharmacist  
6714035                                   Registered Nurse  
6714036  Student in an Organized Health Care Education/...  
6714037                        Speech-Language Pathologist  

[6714038 rows x 3 columns]

In [15]:
specialties = """
    SELECT n.*
        , s.primary_taxonomy
        , s.Classification

    FROM nppes AS n

    JOIN cbsa AS c
        ON c.ZIP = CAST(n.zip_5 AS INTEGER)
    
    LEFT JOIN specialty AS s
        ON s.npi = n.npi
    
    --WHERE entity_type_code = 1 -- only get individuals
"""
# Read the query results into a dataframe
nash_nppes = pd.read_sql(specialties, db)

In [16]:
nash_nppes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38414 entries, 0 to 38413
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   npi                         38414 non-null  int64  
 1   entity_type_code            38414 non-null  float64
 2   provider_organization_name  7574 non-null   object 
 3   provider_last_name          30839 non-null  object 
 4   provider_first_name         30840 non-null  object 
 5   provider_middle_name        19914 non-null  object 
 6   provider_name_prefix        11560 non-null  object 
 7   provider_name_suffix        907 non-null    object 
 8   provider_credential         25323 non-null  object 
 9   address_01                  38414 non-null  object 
 10  address_02                  9957 non-null   object 
 11  city                        38414 non-null  object 
 12  state                       38414 non-null  object 
 13  zip_9                       384

In [17]:
# Take a look at the individual providers who do not have a primary specialty
nash_nppes[(nash_nppes['primary_taxonomy'].isna() == True)
           & (nash_nppes['entity_type_code'] == 1)]#.nunique() # Add this to count them

npi  entity_type_code provider_organization_name  \
258    1114923737               1.0                       None   
333    1801893631               1.0                       None   
1411   1215918321               1.0                       None   
1859   1477532448               1.0                       None   
1987   1346210457               1.0                       None   
2072   1215908058               1.0                       None   
2114   1780656777               1.0                       None   
2133   1245202266               1.0                       None   
2274   1588639561               1.0                       None   
2301   1558337097               1.0                       None   
2308   1073589305               1.0                       None   
2321   1659347557               1.0                       None   
2349   1346217999               1.0                       None   
2356   1992772206               1.0                       None   
2406   1821066341               1.0                       None   
2512   1831158112               1.0                       None   
2603   1598727869               1.0                       None   
2629   1134181720               1.0                       None   
2691   1568425213               1.0                       None   
2750   1336104645               1.0                       None   
2796   1841256112               1.0                       None   
3531   1528006707               1.0                       None   
3537   1386682318               1.0                       None   
3547   1013955707               1.0                       None   
4079   1801832944               1.0                       None   
4129   1437185519               1.0                       None   
4166   1811923246               1.0                       None   
4188   1629005582               1.0                       None   
4197   1962439877               1.0                       None   
4205   1417984964               1.0                       None   
4309   1386673978               1.0                       None   
4462   1053344549               1.0                       None   
4654   1982629952               1.0                       None   
5008   1245253277               1.0                       None   
5083   1316961279               1.0                       None   
5119   1255345385               1.0                       None   
5363   1316955040               1.0                       None   
5746   1194830612               1.0                       None   
5753   1770698425               1.0                       None   
5754   1679688329               1.0                       None   
5833   1013025766               1.0                       None   
5876   1235247339               1.0                       None   
6085   1336241777               1.0                       None   
6125   1720181563               1.0                       None   
6231   1225133267               1.0                       None   
6242   1003912486               1.0                       None   
6306   1891891883               1.0                       None   
6375   1053419382               1.0                       None   
6552   1851491948               1.0                       None   
6661   1992896435               1.0                       None   
7003   1699860056               1.0                       None   
7457   1659469682               1.0                       None   
7519   1538258926               1.0                       None   
7575   1598855934               1.0                       None   
7696   1093897704               1.0                       None   
7927   1457436321               1.0                       None   
8007   1649358581               1.0                       None   
8060   1841378155               1.0                       None   
8061   1982782207               1.0                       None   
8264   1548332950               1.0                   

*There are 104 individual providers who do not have a specialty. At a quick glance, there is no clear connection between these providers - they are in different areas, have different degrees, etc.*

In [18]:
# Take a look at the types of organizations in the dataset
# There are 152 unique classifications plus null values
nash_nppes[nash_nppes['entity_type_code'] == 2].groupby('Classification')['npi'].count().sort_values(ascending = False)

Classification
Clinic/Center                                                                                     981
Pharmacy                                                                                          560
Dentist                                                                                           519
Internal Medicine                                                                                 401
Durable Medical Equipment & Medical Supplies                                                      380
Chiropractor                                                                                      268
Family Medicine                                                                                   261
Specialist                                                                                        228
In Home Supportive Care                                                                           188
Nurse Practitioner                                                 

In [19]:
# Count the number of organizations that do not have a primary specialty
nash_nppes[(nash_nppes['primary_taxonomy'].isna() == True)
           & (nash_nppes['entity_type_code'] == 2)].nunique() # Add this to count them
# There are 85 organizations that do not have a specialty

npi                           85
entity_type_code               1
provider_organization_name    64
provider_last_name             0
provider_first_name            0
provider_middle_name           0
provider_name_prefix           0
provider_name_suffix           0
provider_credential            0
address_01                    81
address_02                    21
city                          20
state                          1
zip_9                         81
zip_5                         34
primary_taxonomy               0
Classification                 0
dtype: int64

### Follow just one NPI through the tables

In [20]:
sample = """
WITH nashville_providers AS (
    SELECT n.*
        , s.Classification AS specialty

    FROM nppes AS n

    JOIN cbsa AS c
        ON c.ZIP = CAST(n.zip_5 AS INTEGER)
    
    LEFT JOIN specialty AS s
        ON s.npi = n.npi
    
    WHERE n.npi = 1215064852 -- Limit to Elizabeth Null only
)
,

referrals_to AS (
    SELECT n.npi
    , n.specialty
    , rto.*
    
    FROM nashville_providers AS n

    JOIN nashville_referrals AS rto
        ON rto.to_npi = n.npi
)
,

referrals_from AS (
    SELECT n.npi
    , n.specialty
    , rfrom.*

    FROM nashville_providers AS n

    JOIN nashville_referrals AS rfrom
        ON rfrom.from_npi = n.npi
)
SELECT *

FROM referrals_to

UNION

SELECT *
FROM referrals_from
;
"""

# Read the query results into a dataframe
liz_null = pd.read_sql(sample, db)

In [21]:
# Take a look at the sample provider dataframe
liz_null

npi               specialty     index    from_npi      to_npi  \
0   1215064852  Psychiatry & Neurology   1777418  1073531042  1215064852   
1   1215064852  Psychiatry & Neurology   2732631  1326130642  1215064852   
2   1215064852  Psychiatry & Neurology  12565235  1467460725  1215064852   
3   1215064852  Psychiatry & Neurology  15967368  1861479545  1215064852   
4   1215064852  Psychiatry & Neurology  25617535  1295726032  1215064852   
5   1215064852  Psychiatry & Neurology  28491271  1578591756  1215064852   
6   1215064852  Psychiatry & Neurology  30072758  1215064852  1104202761   
7   1215064852  Psychiatry & Neurology  30220192  1215064852  1205886264   
8   1215064852  Psychiatry & Neurology  30225822  1205886264  1215064852   
9   1215064852  Psychiatry & Neurology  30342405  1215064852  1295726032   
10  1215064852  Psychiatry & Neurology  30387766  1215064852  1326130642   
11  1215064852  Psychiatry & Neurology  30541653  1215064852  1437194669   
12  1215064852  Psychiatry & Neurology  30581942  1215064852  1467460725   
13  1215064852  Psychiatry & Neurology  30741073  1215064852  1578591756   
14  1215064852  Psychiatry & Neurology  31151047  1215064852  1861479545   
15  1215064852  Psychiatry & Neurology  31234414  1215064852  1922057181   

    patient_count  transaction_count  average_day_wait  std_day_wait  
0              83                 87            46.575        61.971  
1              47                 66            41.848        50.439  
2             261                374            32.687        45.179  
3             247                378            30.696        47.267  
4             203                250            49.400        54.901  
5              50                 75            39.040        48.248  
6             119                151            47.841        53.878  
7             363                637             2.036        14.145  
8             363                613             0.618        10.491  
9             174                224            44.263        50.661  
10             38                 62            24.984        27.715  
11             48                 57            41.281        41.708  
12            233                304            32.648        48.360  
13             40                 59            28.356        38.900  
14            228                338            28.737        45.782  
15             72                 86            38.721        52.519

### Take a look at the total hop-teaming data joined up

In [29]:
nashville_referrals_agg = """
WITH nashville_providers AS (
    SELECT n.*
        , s.Classification AS specialty

    FROM nppes AS n

    JOIN cbsa AS c
        ON c.ZIP = CAST(n.zip_5 AS INTEGER)
    
    LEFT JOIN specialty AS s
        ON s.npi = n.npi
)
,

referrals_to AS (
    SELECT rto.*
    , s.Classification AS from_npi_specialty
    , n.specialty AS to_npi_specialty
    
    FROM nashville_providers AS n

    JOIN nashville_referrals AS rto
        ON rto.to_npi = n.npi
    
    LEFT JOIN specialty AS s
        ON s.npi = rto.from_npi
    
    WHERE n.entity_type_code = 2 -- only get referrals made *to* organizations
)
,

referrals_from AS (
    SELECT rfrom.*
    , n.specialty AS from_npi_specialty
    , s.Classification AS to_npi_specialty

    FROM nashville_providers AS n

    JOIN nashville_referrals AS rfrom
        ON rfrom.from_npi = n.npi
    
    LEFT JOIN specialty AS s
        ON s.npi = rfrom.to_npi

    WHERE n.entity_type_code = 1 -- only get referrals made *by* individual providers
)
,

tot AS (
SELECT *

FROM referrals_to

UNION

SELECT *
FROM referrals_from
)

SELECT COUNT(*) AS tot_rows
    , COUNT(DISTINCT t.from_npi) AS unique_from_npis
    , COUNT(DISTINCT t.to_npi) AS unique_to_npis
    
FROM tot AS t
;
"""
nashville_referrals_agg = pd.read_sql(nashville_referrals_agg, db)

In [30]:
nashville_referrals_agg

tot_rows  unique_from_npis  unique_to_npis
0    174382             28804           20660

*N.B.: The unique number of provider organizations (that is, entity type code 2) is 2.5 times the number of entities within the Nashville CBSA. This would be the case in Nashville area providers are referring to healthcare organizations outside of the Nashville CBSA...*

In [31]:
# Write a query to get every unique from-to NPI pair where
# The from_npi is an individual provider
# The to_npi is an organization
# And one or both of the NPIs are located in the Nashville CBSA

nashville_referrals = """
WITH nashville_providers AS (
    SELECT n.*
        , s.Classification AS specialty

    FROM nppes AS n

    JOIN cbsa AS c
        ON c.ZIP = CAST(n.zip_5 AS INTEGER)
    
    LEFT JOIN specialty AS s
        ON s.npi = n.npi
)
,

referrals_to AS (
    SELECT rto.*
    , s.Classification AS from_npi_specialty
    , n.specialty AS to_npi_specialty
    
    FROM nashville_providers AS n

    JOIN nashville_referrals AS rto
        ON rto.to_npi = n.npi
    
    LEFT JOIN specialty AS s
        ON s.npi = rto.from_npi
    
    WHERE n.entity_type_code = 2 -- only get referrals made *to* organizations
)
,

referrals_from AS (
    SELECT rfrom.*
    , n.specialty AS from_npi_specialty
    , s.Classification AS to_npi_specialty

    FROM nashville_providers AS n

    JOIN nashville_referrals AS rfrom
        ON rfrom.from_npi = n.npi
    
    LEFT JOIN specialty AS s
        ON s.npi = rfrom.to_npi

    WHERE n.entity_type_code = 1 -- only get referrals made *by* individual providers
)

SELECT *

FROM referrals_to

UNION -- use this to ensure that the results are de-duplicated

SELECT *

FROM referrals_from

;
"""
nashville_referrals = pd.read_sql(nashville_referrals, db)

In [32]:
# Take a look at the shape of the resulting dataset
nashville_referrals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 174382 entries, 0 to 174381
Data columns (total 9 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   index               174382 non-null  int64  
 1   from_npi            174382 non-null  int64  
 2   to_npi              174382 non-null  int64  
 3   patient_count       174382 non-null  int64  
 4   transaction_count   174382 non-null  int64  
 5   average_day_wait    174382 non-null  float64
 6   std_day_wait        174382 non-null  float64
 7   from_npi_specialty  173414 non-null  object 
 8   to_npi_specialty    173494 non-null  object 
dtypes: float64(2), int64(5), object(2)
memory usage: 12.0+ MB


In [33]:
# Double check that the counts line up with the aggregate query above
nashville_referrals.from_npi.nunique()

28804

In [34]:
# Double check that the counts line up with the aggregate query above
nashville_referrals.to_npi.nunique()

20660

In [35]:
nashville_referrals.head(100)

index    from_npi      to_npi  patient_count  transaction_count  \
0     705  1013154723  1003028770             32                 50   
1     706  1003963976  1003028770           2535               3945   
2    1139  1003854258  1003052697             31                 55   
3    1713  1043232879  1003076233             76                 78   
4    1755  1023046901  1003077728             61                 67   
5    1757  1003847492  1003077934            170                192   
6    2353  1033109285  1003096553            354                373   
7    3207  1023055126  1003152208            371                759   
8    3208  1003863580  1003152208            509                760   
9    3209  1043297542  1003152208             60                 85   
10   3958  1033109285  1003217522            123                132   
11   4047  1013996222  1003230236            122                126   
12   4059  1003854258  1003230236            119                138   
13   4074  1013179860  1003230236             94                106   
14   4075  1013273044  1003230236             81                 83   
15   4079  1033472592  1003230236            576                712   
16   4080  1043205008  1003230236             72                102   
17   4081  1043206329  1003230236            241                272   
18   4162  1043279821  1003238635             54                 55   
19   4163  1043289945  1003238635             44                 57   
20   4164  1043231343  1003238635             77                 79   
21   4165  1033458518  1003238635             85                130   
22   4166  1043206014  1003238635             88                 92   
23   4167  1003857186  1003238635             46                 52   
24   4168  1043351281  1003238635            268                439   
25   4169  1003078254  1003238635            158                168   
26   4170  1003154352  1003238635             33                 59   
27   4171  1003918475  1003238635             82                133   
28   4172  1023071412  1003238635             36                 60   
29   4173  1023098902  1003238635             49                 86   
30   4174  1013977180  1003238635             65                 94   
31   4175  1023005352  1003238635             48                 56   
32   4176  1033195292  1003238635            109                112   
33   4177  1033192406  1003238635             56                 57   
34   4178  1033197181  1003238635            121                127   
35   4179  1033197678  1003238635             94                188   
36   4180  1033190228  1003238635            369                458   
37   4181  1033217757  1003238635             37                 65   
38   4182  1033249750  1003238635            333                404   
39   4276  1013074442  1003247750             57                256   
40   5675  1003957952  1003808155             47                 68   
41   7872  1033109285  1003834904             48                 57   
42   8683  1013960905  1003848722             60                 65   
43   9540  1023094851  1003860982             73                 89   
44   9657  1043218654  1003862566             95                 96   
45   9658  1043232879  1003862566            377                383   
46   9659  1043300288  1003862566            113                121   
47   9660  1023004041  1003862566             50                 52   
48   9737  1033174982  1003863580            123                146   
49   9738  1033172382  1003863580            162                179   
50   9739  1033175369  1003863580            311                443   
51   9740  1033175807  1003863580            101                108   
52   9741  1033164264  1003863580            362                386   
53   9742  1033131610  1003863580            337                517   
54   9743  1033132212  1003863580            151                194   
55   9744  1033156617  10038

*This dataset has all of the unique from-to NPI pairs where the from_npi is an individual provider, the to_npi is an organization, and one or both of them is located in the Nashville CBSA. The query shows that there are 174,382 unique from-to NPI pairs in Nashville. Once we validate the query, we should be able to use the resulting dataframe as a base for visualizing how different types of providers interact.*

In [36]:
# Write out the complete dataset to a csv to upload into neo4j
nashville_referrals.to_csv('../data/nashville_referrals.csv', index = False)